extensions/amp-cxense-player/0.1/amp-cxense-player.js

```diff
@@ -0,0 +1,189 @@
+      return false;
+    }
+
+    /** @override */
+    buildCallback() {
+      this.element.classList.add('amp-cxense-player');
+
+        /** @private @const {object} */
+      this.cxDefaults_ = JSON.parse(JSON.stringify(cxDefaults));
+        /** @private @const {object} */
+      this.attrs_ = this.getDataAttributes_();
+
+      this.cxDefaults_.embedHost = this.attrs_.debug
+            ? this.cxDefaults_.debugEmbedHost : this.cxDefaults_.distEmbedHost;
+      this.cxDefaults_.embedApp = this.attrs_.debug
+            ? this.cxDefaults_.debugEmbedApp : this.cxDefaults_.distEmbedApp;
+
+        /** @private @const {number} */
+      this.width_ = getLengthNumeral(this.element.getAttribute('width'));
+        /** @private @const {number} */
+      this.height_ = getLengthNumeral(this.element.getAttribute('height'));
+
+    }
+
+    /** @override */
+    createPlaceholderCallback() {
+      const src = this.element.getAttribute('poster')
+            || this.element.getAttribute('data-poster')
+            || this.cxDefaults_.poster;
+
+      const placeholder = this.getWin().document.createElement('div');
+      placeholder.setAttribute('placeholder', '');
+      const image = this.getWin().document.createElement('amp-img');
+      image.setAttribute('src', src);
+      image.setAttribute('layout', 'fill');
+      this.propagateAttributes(['alt'], image);
+      placeholder.appendChild(image);
+      return placeholder;
+    }
+
+    /** @override */
+    layoutCallback() {
+        /** @private @const {Element} */
+      this.iframe_ = this.element.ownerDocument.createElement('iframe');
+
+      this.iframe_.setAttribute('frameborder', '0');
+      this.iframe_.setAttribute('allowfullscreen', 'true');
+
+      this.iframe_.width = this.width_;
+      this.iframe_.height = this.height_;
+
+      this.iframe_.src = this.getIframeSrc_();
+      setStyles(this.iframe_, {display: 'none'});
+
+        /** @private {boolean} */
+      this.mpfReady_ = false;
+
+        /** @private @const {!Promise} */
+      this.playerReadyPromise_ = new Promise(resolve => {
+            /** @private @const {function()} */
+        this.playerReadyResolver_ = iframe => {
+          this.mpfReady = true;
+          resolve(iframe);
+        };
+      });
+      listenForOnce(this.iframe_, 'mpf.ready', () => {
+        this.playerReadyResolver_();
+      });
+      this.element.appendChild(this.iframe_);
+
+      return loadPromise(this.iframe_)
+            // todo: remove this timeout after OVP release
+            // that has the modified frame AMP postMessage support
+            // for a more accurate loading time
+            // this resolves whether mpf.ready fires or not after 2s
+            .then(() => {
+              setTimeout(() => {
+                if (!this.mpfReady_) {
+                  this.playerReadyResolver_(this.iframe_);
+                }
+              }, 2000);
+            })
+            .then(() => this.playerReadyPromise_)
```

Comment:
true, I left them separate to point that the previous `#then` should be completely be removed in the future.